lab_1/main.py

```diff
@@ -1,55 +1,84 @@
-"""
-Lab 1
-A concordance extraction
-"""
-
-
+import re
 def tokenize(text: str) -> list:
     """
-    Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
-    :param text: the initial text
-    :return: a list of lowercased tokens without punctuation
-    e.g. text = 'The weather is sunny, the man is happy.'
-    --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
-    """
-    pass
+        Splits sentences into tokens, converts the tokens into lowercase, removes punctuation
+        :param text: the initial text
+        :return: a list of lowercased tokens without punctuation
+        e.g. text = 'The weather is sunny, the man is happy.'
+        --> ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
+        """
+    if not isinstance(text, str):
+        return []
+
+    else:
+        text = text.lower()
+        text = re.sub(r'[^a-z\s\n]', '', text)  # с помощью рег. выраж. убираю знаки препинания и числительные
+        tokenize_list = text.split()
+
+    return tokenize_list
+
+
+
 
 
 def remove_stop_words(tokens: list, stop_words: list) -> list:
-    """
-    Removes stop words
+    """Removes stop words
     :param tokens: a list of tokens
     :param stop_words: a list of stop words
     :return: a list of tokens without stop words
     e.g. tokens = ['the', 'weather', 'is', 'sunny', 'the', 'man', 'is', 'happy']
     stop_words = ['the', 'is']
     --> ['weather', 'sunny', 'man', 'happy']
     """
-    pass
+    new_tokens = []
+    if isinstance(tokens, list) and isinstance(stop_words, list):#проверяем, списки ли это
+        for token in tokens:
+            if token not in stop_words: #проверяем входит ли слово в список стоп-слов
+                new_tokens.append(token)
+    return new_tokens
+
+
+
 
 
 def calculate_frequencies(tokens: list) -> dict:
     """
-    Calculates frequencies of given tokens
-    :param tokens: a list of tokens without stop words
-    :return: a dictionary with frequencies
-    e.g. tokens = ['weather', 'sunny', 'man', 'happy']
-    --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
-    """
-    pass
+        Calculates frequencies of given tokens
+        :param tokens: a list of tokens without stop words
+        :return: a dictionary with frequencies
+        e.g. tokens = ['weather', 'sunny', 'man', 'happy']
+        --> {'weather': 1, 'sunny': 1, 'man': 1, 'happy': 1}
+        """
+    if not isinstance(tokens, list):
+        return {}
+    dict = {} #создаем словарь
+    if isinstance(tokens[0], str):
```

Comment:
why only the first element?

Reply:
if the first element is a string it doesn't give us a dictionary with the type of None

Reply:
But what if `tokens[0]` is a `str`, but `tokens[1]` is `None`?